### PR TITLE
gh-111089: Add PyUnicode_AsUTF8() to the limited C API

### PR DESCRIPTION
**Pull request by vstinner**
Add PyUnicode_AsUTF8() function to the limited C API.

multiprocessing posixshmem now uses PyUnicode_AsUTF8() instead of PyUnicode_AsUTF8AndSize(): the extension is built with the limited C API. The function now raises an exception if the filename contains an embedded null character instead of truncating silently the filename.

* Issue: gh-111089

----
:books: Documentation preview :books:: https://cpython-previews--111121.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
I just modified the function in PR #111091 to raise an exception if the string contains embedded null characters.

cc @serhiy-storchaka

File: Modules/_multiprocessing/posixshmem.c [hunk 1]
```diff
@@ -44,7 +44,7 @@ _posixshmem_shm_open_impl(PyObject *module, PyObject *path, int flags,
 {
     int fd;
     int async_err = 0;
-    const char *name = PyUnicode_AsUTF8AndSize(path, NULL);
+    const char *name = PyUnicode_AsUTF8(path);
```
Comment (serhiy-storchaka) [inline Modules/_multiprocessing/posixshmem.c:47]: Why not simply use the `str` converter?

Comment (serhiy-storchaka) [reply Modules/_multiprocessing/posixshmem.c:47]: Ah, I see why. `path` is passed to `PyErr_SetFromErrnoWithFilenameObject()`.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by vstinner** [comment]
Merged, thanks for the review @serhiy-storchaka.

**Pull request merged**